### PR TITLE
Ensure services with no operations generate valid Rust

### PR DESCRIPTION
**Pull request by hlbarber**
## Description

- Add test coverage for a service with no operations
- Correct the service builder code to allow for this

### CODE REVIEW

File: codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/ServerServiceGeneratorV2.kt [hunk 1]
```diff
@@ -216,15 +218,14 @@ class ServerServiceGeneratorV2(
             pub fn build(self) -> Result<$serviceName<#{SmithyHttpServer}::routing::Route<$builderBodyGenericTypeName>>, MissingOperationsError>
             {
                 let router = {
-                    use #{SmithyHttpServer}::operation::OperationShape;
+                    ##[allow(unused_mut)]
```
Comment (LukeMathWalker) [inline codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/ServerServiceGeneratorV2.kt:221]: Can we add this conditionally on the fact that we know there are no operations?

File: codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/ServerServiceGeneratorV2.kt [hunk 2]
```diff
@@ -178,18 +178,20 @@ class ServerServiceGeneratorV2(
 
     private fun buildMethod(): Writable = writable {
         val missingOperationsVariableName = "missing_operation_names"
-        val expectMessageVariableName = "unexpected_error_msg"
+        val expectMessage = "this should never panic since we are supposed to check beforehand that a handler has been registered for this operation; please file a bug report under https://github.com/awslabs/smithy-rs/issues"
```
Comment (LukeMathWalker) [inline codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/ServerServiceGeneratorV2.kt:181]: We used a variable there to avoid having multiple copies of the same string in the final binary artefact 🤔

Comment (hlbarber) [reply codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/ServerServiceGeneratorV2.kt:181]: I see, I'd be surprised if it didn't deduplicate them. But I'll put a conditional `#[allow(unused_variables)]` instead.

File: codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/ServerServiceGeneratorV2.kt (hunk 1 above)
Comment (LukeMathWalker) [reply codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/ServerServiceGeneratorV2.kt:221]: Loosening lints in all cases can lead to bugs going undetected (because a legitimate lint doesn't trigger). If it's easy enough to be precise, I think we should be.

Comment (jjant) [reply codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/ServerServiceGeneratorV2.kt:221]: > If it's easy enough to be precise, I think we should be.

+1 to this. I don't think we should bend over backwards to have lints be as tight as possible, but if it's easy I'd go for it.

File: codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/ServerServiceGeneratorV2.kt [hunk 3]
```diff
@@ -184,12 +184,14 @@ class ServerServiceGeneratorV2(
             for (operationShape in operations) {
                 val fieldName = builderFieldNames[operationShape]!!
                 val operationZstTypeName = operationStructNames[operationShape]!!
-                rust(
+                rustTemplate(
                     """
                     if self.$fieldName.is_none() {
+                        use #{SmithyHttpServer}::operation::OperationShape;
```
Comment (jjant) [inline codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/ServerServiceGeneratorV2.kt:190]: Why was this needed? 🤔

File: codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/ServerServiceGeneratorV2.kt [hunk 4]
```diff
@@ -205,6 +207,13 @@ class ServerServiceGeneratorV2(
             }
         }
 
+        var missingOperationsUnusedMut = ""
+        var unusedExpected = ""
+        if (operations.isEmpty()) {
+            missingOperationsUnusedMut = "##[allow(unused_mut)]"
+            unusedExpected = "##[allow(unused_variables)]"
+        }
+
```
Comment (jjant) [inline codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/ServerServiceGeneratorV2.kt:216]: I think you can use `Attribute.AllowUnusedMut/AllowUnusedVariables.render(writer)` for these.

Comment (hlbarber) [reply codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/ServerServiceGeneratorV2.kt]: https://github.com/awslabs/smithy-rs/pull/2351/commits/e0a4d8316c92fa78d12c420ae7f58d89255585e1

**Review by LukeMathWalker — APPROVED** [review]

**Comment by hlbarber** [comment]
Following the decision made in https://github.com/awslabs/smithy-rs/pull/1369 we should NOT generate code for this Smithy class of models and instead throw an exception.
